### PR TITLE
[hwasan] Classify stack overflow, and use after scope

### PR DESCRIPTION
**Pull request by vitalybuka**
We can't distinguish UAR and UAS, but by definition
UAR is already UAS.

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 1]
```diff
@@ -221,29 +221,55 @@ static void PrintStackAllocations(const StackAllocationsRingBuffer *sa,
...
+        Printf("%p is located %zd bytes %s a %zd-byte region [%p,%p)\n",
+               untagged_addr, offset, whence, local_end - local_beg, local_beg,
+               local_end);
+        Printf("%s", d.Allocation());
         StackTracePrinter::GetOrInit()->RenderSourceLocation(
             &location, local.decl_file, local.decl_line, 0,
             common_flags()->symbolize_vs_style,
             common_flags()->strip_path_prefix);
-        Printf("  %s in %s %s\n", local.name, local.function_name,
+        Printf("declared as %s in %s %s\n", local.name, local.function_name,
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:269]: I would split out this change. It's a kind of unrelated behaviour change.

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 2]
```diff
@@ -221,29 +221,55 @@ static void PrintStackAllocations(const StackAllocationsRingBuffer *sa,
       for (LocalInfo &local : frame.locals) {
         if (!local.has_frame_offset || !local.has_size || !local.has_tag_offset)
           continue;
+        if (!(local.name && internal_strlen(local.name)) &&
+            !(local.function_name && internal_strlen(local.name)) &&
+            !local.decl_file)
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:226]: why don't we check this for zero size? please add in comment

File: compiler-rt/test/hwasan/TestCases/stack-underflow.c [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  char c[64];
+  char *volatile p = c;
+  p[-2] = 0;
+}
+
+int main() {
+  buggy();
+  // CHECK: WRITE of size 1 at
+  // CHECK: #0 {{.*}} in buggy{{.*}}stack-underflow.c:[[@LINE-6]]
+  // CHECK: Cause: stack tag-mismatch
```
Comment (fmayer) [inline compiler-rt/test/hwasan/TestCases/stack-underflow.c:16]: I am not sure about the output format. It has two `Cause:` clauses that are different to what this used to be. The second `Cause` is a more detailed description of the first, not an entirely different cause.

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 2 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp]: will check

File: compiler-rt/test/hwasan/TestCases/stack-underflow.c (hunk 3 above)
Comment (vitalybuka) [reply compiler-rt/test/hwasan/TestCases/stack-underflow.c:16]: The first one is true, the second one is "Potentially"

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 4]
```diff
@@ -221,29 +221,55 @@ static void PrintStackAllocations(const StackAllocationsRingBuffer *sa,
...
+        Decorator d;
+        Printf("%s", d.Error());
+        Printf("Cause: %s\n", cause);
+        Printf("%s", d.Default());
+        Printf("%s", d.Location());
+        Printf("%p is located %zd bytes %s a %zd-byte region [%p,%p)\n",
+               untagged_addr, offset, whence, local_end - local_beg, local_beg,
+               local_end);
+        Printf("%s", d.Allocation());
         StackTracePrinter::GetOrInit()->RenderSourceLocation(
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:265]: FYI the offline symbolizer has this output format

```
        self.print('')
        self.print('Potentially referenced stack object:')
        self.print('  %d bytes inside a variable "%s" in stack frame of function "%s"' % (obj_offset, local[2], local[0]))
        self.print('  at %s' % (local[1],))
```

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 5]
```diff
@@ -221,29 +221,55 @@ static void PrintStackAllocations(const StackAllocationsRingBuffer *sa,
...
-        // Calculate the offset from the object address to the faulting
-        // address. Because we only store bits 4-19 of FP (bits 0-3 are
-        // guaranteed to be zero), the calculation is performed mod 2^20 and may
-        // harmlessly underflow if the address mod 2^20 is below the object
-        // address.
-        uptr obj_offset =
-            (untagged_addr - fp - local.frame_offset) & (kRecordFPModulus - 1);
-        if (obj_offset >= local.size)
-          continue;
+        uptr local_beg = (fp + local.frame_offset) |
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:231]: I am confused by this. Could you add a comment as on the LHS? Why isn't the `local_beg` not just `fp + local.frame_offset`?

**Review by fmayer — COMMENTED** [review]
Remove this comment?

Line 780

```
    // TODO(fmayer): figure out how to distinguish use-after-return and
    // stack-buffer-overflow.
```

**Comment by vitalybuka** [comment]
> Remove this comment?
> 
> Line 780
> 
> ```
>     // TODO(fmayer): figure out how to distinguish use-after-return and
>     // stack-buffer-overflow.
> ```

done

**Review by fmayer — APPROVED** [review]

**Comment by antmox** [comment]
Hi, looks like patch broke 2 aarch64 bots:
https://lab.llvm.org/buildbot/#/builders/185/builds/5675
https://lab.llvm.org/buildbot/#/builders/179/builds/8880
Could you please look at this?

**Pull request merged**